nemo/collections/asr/parts/utils/nmesc_clustering.py

```diff
@@ -154,218 +389,282 @@ def get_argmin_mat(uniq_scale_dict):
     """
     laplacian = getLaplacian(affinity_mat)
     lambdas, _ = eigDecompose(laplacian, is_cuda)
-    lambdas = np.sort(lambdas)
-    lambda_gap_list = getLamdaGaplist(lambdas)
-    num_of_spk = np.argmax(lambda_gap_list[: min(max_num_speaker, len(lambda_gap_list))]) + 1
-    return num_of_spk, lambdas, lambda_gap_list
+    lambdas = torch.sort(lambdas)[0]
+    lambda_gap = getLamdaGaplist(lambdas)
+    num_of_spk = torch.argmax(lambda_gap[: min(max_num_speaker, lambda_gap.shape[0])]) + 1
+    return num_of_spk, lambdas, lambda_gap
 
 
-class _SpectralClustering:
-    def __init__(self, n_clusters=8, random_state=0, n_init=10, p_value=10, n_jobs=None, cuda=False):
+@torch.jit.script
+class SpectralClustering:
+    def __init__(
+        self,
+        n_clusters: int = 8,
+        random_state: int = 0,
+        n_init: int = 10,
+        p_value: int = 10,
+        cuda: bool = False,
+        device: torch.device = torch.device('cpu'),
+    ):
         self.n_clusters = n_clusters
         self.random_state = random_state
         self.n_init = n_init
         self.p_value = p_value
-        self.affinity_matrix_ = None
         self.cuda = cuda
+        self.device = device
 
     def predict(self, X):
         if X.shape[0] != X.shape[1]:
             raise ValueError("The affinity matrix is not a square matrix.")
-
-        self.affinity_matrix_ = X
-        labels = self.clusterSpectralEmbeddings(self.affinity_matrix_, n_init=self.n_init, cuda=self.cuda)
+        labels = self.clusterSpectralEmbeddings(X, n_init=self.n_init, cuda=self.cuda, device=self.device)
         return labels
 
-    def clusterSpectralEmbeddings(self, affinity, n_init=10, cuda=False):
+    def clusterSpectralEmbeddings(
+        self, affinity, n_init: int = 10, cuda: bool = False, device: torch.device = torch.device('cpu')
+    ):
         spectral_emb = self.getSpectralEmbeddings(affinity, n_spks=self.n_clusters, drop_first=False, cuda=cuda)
-        _, labels, _ = k_means(spectral_emb, self.n_clusters, random_state=self.random_state, n_init=n_init)
+        labels = kmeans_torch(X=spectral_emb, num_clusters=self.n_clusters, device=device)
         return labels
 
-    def getSpectralEmbeddings(self, affinity_mat, n_spks=8, drop_first=True, cuda=False):
-        if not isGraphFullyConnected(affinity_mat):
-            logging.warning("Graph is not fully connected and the clustering result might not be accurate.")
-
+    def getSpectralEmbeddings(
+        self, affinity_mat: torch.Tensor, n_spks: int = 8, drop_first: bool = True, cuda: bool = False
+    ):
         laplacian = getLaplacian(affinity_mat)
         lambdas_, diffusion_map_ = eigDecompose(laplacian, cuda)
         diffusion_map = diffusion_map_[:, :n_spks]
-        embedding = diffusion_map.T[n_spks::-1]
+        inv_idx = torch.arange(diffusion_map.size(1) - 1, -1, -1).long()
+        embedding = diffusion_map.T[inv_idx, :]
         return embedding[:n_spks].T
 
 
+@torch.jit.script
```

Comment:
oh I didn't know we can @torch.jit.script the whole class. Does it mean all functions within this class is torch-jit-scriptable?